### ISSUE TITLE
reserved_concurrent_executions  option of `aws_lambda_function` gives error message

### ISSUE THREAD

**Issue by maanenh** [terraform-providers/terraform-provider-aws#3563]
The `reserved_concurrent_executions`  option of `aws_lambda_function` does not work.

### Terraform Version
* Terraform v0.11.3
* provider.archive v1.0.0
* provider.aws v1.10.0

### Affected Resource(s)
aws_lambda_function

### Terraform Configuration Files
```hcl
data "archive_file" "my_lambda" {
  type = "zip"
  source_file = "./myLambda.js"
  output_path = "myLambda.zip"
}

resource "aws_lambda_function" "my_lambda" {
  filename         = "${data.archive_file.my_lambda.output_path}"
  function_name    = "concurrentLambdaIssue"
  role             = "${var.role_arn}"
  handler          = "myLambda.handler"
  runtime          = "nodejs6.10"
  source_code_hash = "${base64sha256(file("${data.archive_file.my_lambda.output_path}"))}"
  
  reserved_concurrent_executions = 2
}
```

### Debug Output

First execution:
* aws_lambda_function.my_lambda: Error setting concurrency for Lambda concurrentLambdaIssue: ResourceNotFoundException: An error occurred and the request cannot be processed.
        status code: 404, request id: xx

When you run terraform apply again the messages is:
* aws_lambda_function.my_lambda: Error creating Lambda function: ResourceConflictException: Function already exist: concurrentLambdaIssue
        status code: 409, request id: xx

### Panic Output
n/a

### Expected Behavior
Function created with reserved_concurrent_executions  set.

### Actual Behavior
See error messages.

### Steps to Reproduce
1. `terraform apply`

### Important Factoids
* On first execution the message is: ResourceNotFoundException.  When you execute `terraform apply` again the messages is: `Function already exist`.
* It seems the function is created during the first execution, but the state-files is not updated with this information.
* When you look with the AWS console, the function is indeed created, but without `reserved_concurrent_executions`

**Comment by bflad** [comment]
Hi @maanenh 👋  thank you very much for this bug report and sorry you're running into trouble here.

We actually acceptance test this regularly without issue but looks like there could be some latent eventual consistency within the AWS Lambda service. There's two fixes we can apply here to help the situation:
* Calling `d.SetId()` before we call `PutFunctionConcurrency` so any errors returned by that won't prevent Terraform from seeing the existing resource (it'll just show a plan to update the concurrency again if it errored on creation)
* Retrying for up to a minute on the `ResourceNotFound` error with `PutFunctionConcurrency` which should catch eventual consistency issues in most cases

I have a PR ready to go for this, which I'm acceptance testing now!

**Comment by bflad** [comment]
Both the above fixes have been implemented into master and will be released with v1.11.0 of the AWS provider, likely at the end of this week.

**Comment by maanenh** [comment]
Thanks for fixing this issue.

**Comment by bflad** [comment]
This has been released in [version 1.11.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1110-march-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
